doc/in-progress/custom-upgrade-policy.md

```diff
@@ -0,0 +1,1000 @@
+(identified by `cap: ID`) with the ID of the latest package in its
+family (`package: ID`).
+
+Once an `UpgradeReceipt` is created, it must be used to update its
+`UpgradeCap` within the same transaction (it cannot be stored for
+later, dropped, or burned), or the transaction will fail.
+
+## Isolating Policies
+
+When writing custom upgrade policies, prefer: 
+
+- separating them into their own package, (i.e. not co-located with
+  the code they govern the upgradeability of)
+- making that package immutable (not upgradeable),
+- locking in the `UpgradeCap`, so that a package's upgrade policy
+  cannot be made less restrictive, once one is chosen.
+
+These best practices help uphold **informed user consent** and
+**bounded risk** by making it clear what a package's upgrade policy is
+at the moment a user locks value into it, and ensuring that the policy
+does not evolve to be more permissive with time, without the package
+user realising and choosing to accept the new terms.
+
+## Example: "Day of the Week" Upgrade Policy
+
+We will put everything into practice by writing a toy upgrade policy that
+only authorizes upgrades on a particular day of the week (of the
+package creator's choosing).
+
+Start by creating a new move package for the upgrade policy:
+
+```
+$ sui move new policy
+```
+
+Add a source file, at `policy/sources/day_of_week.move`, where we
+define the object type for our custom upgrade policy, and a
+constructor:
+
+```rust
+module policy::day_of_week {
+    use sui::object::{Self, UID};
+    use sui::package;
+    use sui::tx_context::TxContext;
+
+    struct UpgradeCap has key, store {
+        id: UID,
+        cap: package::UpgradeCap,
+        day: u8,
+    }
+
+    /// Day is not a week day (number in range 0 <= day < 7).
+    const ENotWeekDay: u64 = 1;
+
+    public fun new_policy(
+        cap: package::UpgradeCap,
+        day: u8,
+        ctx: &mut TxContext,
+    ): UpgradeCap {
+        assert!(day < 7, ENotWeekDay);
+        UpgradeCap { id: object::new(ctx), cap, day }
+    }
+}
+```
+
+We then need to add a function to authorize an upgrade, if we are on
+the correct day of the week:
+
+```rust
+module policy::day_of_week {
+    use sui::package;
+    use sui::tx_context::{Self, TxContext};
+
+    /// Request to authorize upgrade on the wrong day of the week.
+    const ENotAllowedDay: u64 = 2;
+
+    const MS_IN_DAY: u64 = 24 * 60 * 60 * 1000;
+
+    public fun authorize_upgrade(
+        cap: &mut UpgradeCap,
+        policy: u8,
+        digest: vector<u8>,
+        ctx: &TxContext,
+    ): package::UpgradeTicket {
+        assert!(week_day(ctx) == cap.day, ENotAllowedDay);
+        package::authorize_upgrade(&mut cap.cap, policy, digest)
+    }
+
+    fun week_day(ctx: &TxContext): u8 {
+        let days_since_unix_epoch = 
+            tx_context::epoch_timestamp_ms(clock) / MS_IN_DAY;
+        // The unix epoch (1st Jan 1970 was a Thursday)
+        ((days_since_unix_epoch + 3) % 7 as u8)
```

Comment:
Checking my math, we get `+ 3` from `0` being Monday, ... `3` being Thursday? If so might be worth putting in comments 😄

Reply:
Yeah that's right -- let me make this comment a bit less terse!